src/selectors/columnData.js

```diff
@@ -0,0 +1,148 @@
+import shallowEqualSelector from 'shallowEqualSelector';
+import pick from 'lodash/pick';
+
+/**
+ * Converts React column / column group elements into props and cell rendering templates
+ */
+function convertColumnElementsToData(childComponents) {
+  const children = [];
+  React.Children.forEach(childComponents, (child) => {
+    if (child == null) {
+      return;
+    }
+    invariant(child.type.__TableColumnGroup__ || child.type.__TableColumn__,
+      'child type should be <FixedDataTableColumn /> or <FixedDataTableColumnGroup />');
+
+    children.push(child);
+  });
+
+  const elementTemplates = getDefaultElementTemplates();
+
+  const columnProps = [];
+  const hasGroupHeader = children.length && children[0].type.__TableColumnGroup__;
+  if (hasGroupHeader) {
+    const columnGroupProps = map(children, extractPropsFromElement);
+    forEach(children, (columnGroupElement, index) => {
+      elementTemplates.groupHeader.push(columnGroupElement.props.header);
+
+      React.Children.forEach(columnGroupElement.props.children, (child) => {
+        const column = extractPropsFromElement(child);
+        column.groupIdx = index;
+        columnProps.push(column);
+        extractTemplatesFromElement(elementTemplates, child);
+      });
+    });
+
+    return {
+      columnGroupProps,
+      columnProps,
+      elementTemplates,
+      useGroupHeader: true,
+    };
+  }
+
+  // Use a default column group
+  forEach(children, (child) => {
+    columnProps.push(extractPropsFromElement(child));
+    extractTemplatesFromElement(elementTemplates, child);
+  });
+  return {
+    columnGroupProps: [],
+    columnProps,
+    elementTemplates,
+    useGroupHeader: false,
+  };
+}
+
+function fixColumnPropsWithColumnGroupProps(columnProps, columnGroupProps) {
+  const columnGroups = [];
+  let columnsWithDefinedGroupIndex = 0;
+
+  forEach(columnProps, ({ groupIdx }) => {
+    if (groupIdx !== undefined && groupIdx !== null) {
+      columnGroups[groupIdx] = true;
+      columnsWithDefinedGroupIndex++;
+    }
+  });
+
+  invariant(columnGroups.length === (columnGroupProps || []).length,
+    'column group data must fully and only contain every column groups');
+
+  invariant(columnsWithDefinedGroupIndex === columnProps.length,
+    'group index if specified must be given for every column');
+
+  forEach(columnProps, (columnProp) => {
+    columnProp.fixed = columnGroupProps[columnProp.groupIdx].fixed;
+    columnProp.fixedRight = columnGroupProps[columnProp.groupIdx].fixedRight;
+  });
+}
+
+/**
+ * Uses columnData to extract column props and cell rendering templates
+ *
+ * @param {!Array.<columnDefinition>} columnData
+ * @param {!Array.<columnDefinition>} columnGroupData
+ * @return {{
+ *   columnGroupProps: !Array.<columnDefinition>,
+ *   columnProps: !Array.<columnDefinition>,
+ *   elementTemplates: !Object.<string, Array>,
+ *   useGroupHeader: boolean,
+ * }}
+ */
+function getColumnPropsFromData(columnData, columnGroupData) {
```

Comment:
this is the starting point for the new API
steps are:
* collect column props from data
* collect column group props from data
* collect templates from data
* fix column props (explained above)